### PR TITLE
PE-3991: update quote every 5 minutes

### PR DESCRIPTION
**Pull request by thiagocarvalhodev**
…plement a listener to update when quote should be updated

--- Releases ---
Android release: https://appdistribution.firebase.google.com/testerapps/1:305132849030:android:6cf0cd5ec064fad3ffce07/releases/7eq6f17alj5kg
iOS release: https://appdistribution.firebase.google.com/testerapps/1:305132849030:ios:06ea33a95a2e0d42ffce07/releases/6u1q1m0tg6j10

### CODE REVIEW

File: lib/turbo/turbo.dart [hunk 1]
```diff
@@ -9,88 +9,176 @@ import 'package:ardrive/utils/logger/logger.dart';
...
+  }
+
+  late Timer _priceEstimateTimer;
+
+  final StreamController<PriceEstimate> _priceEstimateController =
+      StreamController.broadcast();
+
+  void _startOnPriceEstimateChange() {
+    _priceEstimateTimer =
+        Timer.periodic(const Duration(minutes: 5), (timer) async {
```
Comment (karlprieb) [inline lib/turbo/turbo.dart:67]: create a const `expirationTime` (or something similar) for 5 minutes and use it everywhere

File: lib/turbo/turbo.dart [hunk 2]
```diff
@@ -9,88 +9,176 @@ import 'package:ardrive/utils/logger/logger.dart';
...
-  FileSizeUnit _currentDataUnit = FileSizeUnit.gigabytes;
+  late final DateTime _initialSessionTime;
+  late Timer _sessionExpirationTimer;
 
-  PriceEstimate _priceEstimate = PriceEstimate(
-    credits: BigInt.from(0),
-    priceInCurrency: 0,
-    estimatedStorage: 0,
-  );
+  // Constructor
```
Comment (karlprieb) [inline lib/turbo/turbo.dart:93]: remove comment

File: test/turbo/turbo_test.dart [hunk 3]
```diff
@@ -0,0 +1,572 @@
...
+                currentCurrency: 'usd',
+              )).thenAnswer((_) async => mockPriceEstimate2);
+
+          async.elapse(const Duration(minutes: 5));
+
+          expect(priceEstimate, mockPriceEstimate2);
+        });
+      });
+      test(
+          'priceEstimate updates when price quote expires even if the user doenst select any amount',
```
Comment (karlprieb) [inline test/turbo/turbo_test.dart:228]: add test for when the currency and the currency amount changes

File: test/turbo/turbo_test.dart [hunk 4]
```diff
@@ -0,0 +1,572 @@
...
+
+      final cost = await costCalculator.getCostOfOneGB();
+
+      expect(cost, equals(mockCost));
+      verify(() => mockPaymentService.getPriceForBytes(byteSize: byteSize))
+          .called(1);
+    });
+
+    test(
+        'getCostOfOneGB returns cached cost if fetched less than 10 minutes ago',
```
Comment (karlprieb) [inline test/turbo/turbo_test.dart:329]: 5 minutes

File: test/turbo/turbo_test.dart [hunk 5]
```diff
@@ -0,0 +1,572 @@
...
+            await turboPriceEstimator.computeStorageEstimateForCredits(
+          credits: expectedCredits,
+          outputDataUnit: outputDataUnit,
+        );
+
+        // Verify the result
+        expect(estimatedStorage, expectedEstimatedStorage);
+      });
+
+      test('should return correct estimate using MiB', () async {
```
Comment (karlprieb) [inline test/turbo/turbo_test.dart:506]: test with bytes

File: lib/turbo/turbo.dart [hunk 6]
```diff
@@ -110,62 +198,28 @@ class Turbo {
...
-    _currentCurrency = currentCurrency;
-    _currentDataUnit = currentDataUnit;
-
-    _priceEstimate = await computePriceEstimate(
-      currentAmount: currentAmount,
-      currentCurrency: currentCurrency,
-      currentDataUnit: currentDataUnit,
-    );
 
     return _priceEstimate;
```
[the remark this answers is not in the document]
Comment (thiagocarvalhodev) [reply lib/turbo/turbo.dart:207]: It's implemented on here: https://github.com/ardriveapp/ardrive-web/blob/4ed48bb79238d40cc19e554697714a1b8e8b5521/lib/turbo/turbo.dart#L51

File: lib/turbo/turbo.dart (hunk 2 above)
Comment (thiagocarvalhodev) [reply lib/turbo/turbo.dart]: https://github.com/ardriveapp/ardrive-web/pull/1135/commits/ab9197399559f680ad43ea6d6e88f65a26c5e51d

File: lib/turbo/turbo.dart (hunk 1 above)
Comment (thiagocarvalhodev) [reply lib/turbo/turbo.dart]: https://github.com/ardriveapp/ardrive-web/pull/1135/commits/ab9197399559f680ad43ea6d6e88f65a26c5e51d

File: lib/turbo/turbo.dart [hunk 7]
```diff
@@ -9,88 +9,176 @@ import 'package:ardrive/utils/logger/logger.dart';
...
+    if (!forceGet &&
+        _costOfOneGb != null &&
+        _lastCostOfOneGbFetchTime != null) {
+      final difference = currentTime.difference(_lastCostOfOneGbFetchTime!);
       if (difference.inMinutes < 10) {
         return _costOfOneGb!;
       }
     }
 
     _costOfOneGb =
```
[the remark this answers is not in the document]
Comment (thiagocarvalhodev) [reply lib/turbo/turbo.dart:150]: I'd say to keep using the same method as we can call that endpoint many times.

File: lib/turbo/turbo.dart [hunk 8]
```diff
@@ -9,88 +9,176 @@ import 'package:ardrive/utils/logger/logger.dart';
...
   Future<BigInt> getCostOfOneGB({bool forceGet = false}) async {
     final currentTime = DateTime.now();
 
-    if (!forceGet && _costOfOneGb != null && _lastFetchTime != null) {
-      final difference = currentTime.difference(_lastFetchTime!);
+    if (!forceGet &&
+        _costOfOneGb != null &&
+        _lastCostOfOneGbFetchTime != null) {
+      final difference = currentTime.difference(_lastCostOfOneGbFetchTime!);
       if (difference.inMinutes < 10) {
```
[the remark this answers is not in the document]
Comment (thiagocarvalhodev) [reply lib/turbo/turbo.dart]: https://github.com/ardriveapp/ardrive-web/pull/1135/commits/ab9197399559f680ad43ea6d6e88f65a26c5e51d

**Pull request merged**